File: 2015/lua/3/main-part-1.lua
local Position = {
    x = 0,
    y = 0
}

function move(Position, direction)
    if direction == "^" then
        Position.y = Position.y + 1
    elseif direction == "v" then
        Position.y = Position.y - 1
    elseif direction == "<" then
        Position.x = Position.x - 1
    elseif direction == ">" then
        Position.x = Position.x + 1
    end
end

function main()
    if #arg < 1 then
        print("Usage: lua main.lua <input_file>")
        os.exit(1)
    end

    local filename = arg[1]
    local file = io.open(filename, "r")
    if (file == nil) then
        print("Error: Could not open file " .. filename)
        os.exit(1)
    end

    local content = file:read("*all")
    file:close()

    -- hash table to store the number of houses visited
    local houses = {}
    local key = Position.x .. "," .. Position.y
    houses[key] = 1

    -- loop each line
    for line in content:gmatch("[^\r\n]+") do
        -- loop each character
        for c in line:gmatch(".") do
            move(Position, c)
            key = Position.x .. "," .. Position.y
            if houses[key] == nil then
                houses[key] = 1
            else
                houses[key] = houses[key] + 1
            end
        end
    end

    local count = 0
    for k, v in pairs(houses) do
        count = count + 1
    end

    print("Part 1: " .. count)
end

main()

File: 2015/php/3/main-part-1.php
<?php

class Position {
    public $x;
    public $y;

    public function __construct($x, $y) {
        $this->x = $x;
        $this->y = $y;
    }

    public function move($direction) {
        switch($direction) {
            case '^':
                $this->y--;
                break;
            case 'v':
                $this->y++;
                break;
            case '<':
                $this->x--;
                break;
            case '>':
                $this->x++;
                break;
        }
    }
}

function main() {
    // get command line arguments
    $args = $_SERVER['argv'];
    if(count($args) < 2) {
        echo "Usage: php main.php <input_file>\n";
        return;
    }

    $filename = $args[1];

    // read input file
    $file = fopen($filename, "r");
    if($file === false) {
        echo "Error: Unable to open file $filename\n";
        return;
    }

    $input = [];
    while(!feof($file)) {
        $line = trim(fgets($file));
        if($line !== "") {
            $input[] = $line;
        }
    }

    fclose($file);

    // create a hashmap to store the number of houses visited
    $houses = [];
    $pos = new Position(0, 0);
    $houses[$pos->x][$pos->y] = 1;

    // loop each line of input
    foreach($input as $line) {
        for($i = 0; $i < strlen($line); $i++) {
            $pos->move($line[$i]);
            if(!isset($houses[$pos->x])) {
                $houses[$pos->x] = [];
            }
            if(!isset($houses[$pos->x][$pos->y])) {
                $houses[$pos->x][$pos->y] = 1;
            } else {
                $houses[$pos->x][$pos->y]++;
            }
        }
    }

    // count the number of houses visited
    $count = 0;
    foreach($houses as $x => $row) {
        $count += count($row);
    }

    echo "Part 1: $count\n";
}

main();